mcs/class/corlib/System.Reflection.Emit/ModuleBuilder.cs

```diff
@@ -376,44 +369,33 @@ public override Type GetType (string className, bool throwOnError, bool ignoreCa
 			if (className.Length == 0)
 				throw new ArgumentException ("className");
 
-			int subt;
-			string orig = className;
-			string modifiers;
 			TypeBuilder result = null;
 
 			if (types == null && throwOnError)
 				throw new TypeLoadException (className);
 
-			subt = className.IndexOfAny (type_modifiers);
-			if (subt >= 0) {
-				modifiers = className.Substring (subt);
-				className = className.Substring (0, subt);
-			} else
-				modifiers = null;
+			TypeSpec ts = TypeSpec.Parse (className);
 
 			if (!ignoreCase) {
-				result =  name_cache [className] as TypeBuilder;
+				var displayNestedName = ts.WithoutModifiers().TypeName;
+				name_cache.TryGetValue (displayNestedName, out result);
 			} else {
-				subt = className.IndexOf ('+');
-				if (subt < 0) {
-					if (types != null)
-						result = search_in_array (types, num_types,  className);
-				} else {
-					string pname, rname;
-					pname = className.Substring (0, subt);
-					rname = className.Substring (subt + 1);
-					result = search_in_array (types, num_types, pname);
-					if (result != null)
-						result = GetMaybeNested (result, rname);
+				if (types != null)
+					result = search_in_array (types, num_types,  ts.Name);
+				if (!ts.IsNested && result != null) {
+					result = GetMaybeNested (result, ts.Nested);
 				}
 			}
 			if ((result == null) && throwOnError)
-				throw new TypeLoadException (orig);
-			if (result != null && (modifiers != null)) {
-				Type mt = create_modified_type (result, modifiers);
-				result = mt as TypeBuilder;
-				if (result == null)
-					return mt;
+				throw new TypeLoadException (className);
+			if (result != null && (ts.HasModifiers || ts.IsByRef)) {
+				Type mt = result.CreateType();
```

Comment:
Okay, I got this pretty wrong. `create_modified_type()` did not cause the `TypeBuilder` to be created.  It did something pretty weird - it grabbed the (not finished) `MonoType` of the typebuilder (setup by `TypeBuilder.setup_internal_class`) and started creating pointer and array types from that `MonoType` and then turning the result of that process into a `MonoReflectionType` (which gets cached by the monodomain so in the future callers get the same reflection type back).

So you got this weird situation where on the managed side you'd get a perfectly usable `MonoType` for the pointer or the array, but if you called `GetElementType()` on that pointer or array, you'd get back the `Type` that the `TypeBuilder` would have created, except that it's in a weird not-created state.